server/api/authentication/authenticator.go

```diff
@@ -0,0 +1,35 @@
+package authentication
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"github.com/tobocop/go-teleport-directory-browser/api/user"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Authenticator interface {
+	Authenticate(username string, password string) (bool, error)
+}
+
+// I'd provide this via a credential manager (aws secrets, credhub, etc) or via an environment variable that was populated through more secure means
+var hmacKey = []byte("8cf7a749-bf77-42ad-abc4-7cf110872bc4")
+
+type staticCredentialsAuthenticator struct {
+	UserStore user.Store
+}
+
+func (sa *staticCredentialsAuthenticator) Authenticate(username string, password string) (bool, error) {
+	mac := hmac.New(sha512.New, hmacKey)
+	mac.Write([]byte(password))
+	hash := hex.EncodeToString(mac.Sum(nil))
+
+	hashed, err := bcrypt.GenerateFromPassword([]byte(hash), 10)
+	if err != nil {
+		return false, err
+	}
+
+	found, password := sa.UserStore.GetUserPassword(username)
+
+	return found && bcrypt.CompareHashAndPassword(hashed, []byte(password)) == nil, nil
```

Comment:
seems that a password is stored in plain text then, no?

Reply:
As far as I could tell, the built in golang bcrypt functions don't allow you to set the salt explicitly and it's generated uniquely for a given hash. The comparison function requires a plaintext version of the password and since it was the algo we decided on in the design docs I stuck with it.

I tried to mitigate this by storing the passwords as hmac-sha's instead of the raw text of a given password.

Reply:
The bcrypt hash is not just the hash, it's the salt + hash. When you do `bcrypt.CompareHashAndPassword` it just extracts the salt from the bcrypt hash, this thing: $2a$10$J2dMyKwK4188qqjTnk/vkurcdEUMRt/Bq0zoa6isxdsxHWKIPFoyu. Then uses it to recompute the hash and compare.

Reply:
Yup, totally. Code should be updated to reflect an appropriate implementation now.